### PR TITLE
Revert "Revert "Use dictionary lookup while resolving assembly refere…

### PR DESCRIPTION
**Pull request by ToddGrun**
…nces (#66430)""

This reverts commit 78edab8532e26d91ae3288f02a346ac6adc37b61.

### CODE REVIEW

**Comment by ToddGrun** [comment]
Removing draft status, not ready for review until test insertion PR numbers come through

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs [hunk 1]
```diff
@@ -925,15 +925,19 @@ private static void AddModule(PEModule module, int referenceIndex, ResolvedRefer
...
-                AssemblyIdentity definition = definitions[i].Identity;
+                // Skip assembly being built for now; it will be considered at the very end
+                if (definitionIndex == 0)
+                {
+                    continue;
+                }
+
+                AssemblyIdentity definition = definitionData.Identity;
 
                 if (!AssemblyIdentityComparer.SimpleNameComparer.Equals(reference.Name, definition.Name))
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs:942]: >if (!AssemblyIdentityComparer.SimpleNameComparer.Equals(reference.Name, definition.Name)) [](http://example.com/codeflow?start=16&length=89)

Is this still necessary? If not, consider removing or converting to an assert.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs (hunk 1 above)
Comment (ToddGrun) [reply src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs:942]: From what I see, it doesn't look to be necessary. I think I would feel safer leaving it in as an assert. Would you rather have me assert here, or assert outside the loop by exposing the MultiDictionary's comparer?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

**Comment by ToddGrun** [comment]
Other PR was taken instead of this one

**Pull request closed without merging**